R/utils.R

```diff
@@ -191,7 +191,19 @@ check_filter_datanames <- function(filters, datanames) {
 teal_data_datanames <- function(data) {
   checkmate::assert_class(data, "teal_data")
   if (length(teal.data::datanames(data))) {
-    teal.data::datanames(data)
+    datanames <- teal.data::datanames(data)
+    joinkeys <- teal.data::join_keys(data)
+    if (length(join_keys()) > 0) {
+      child_parent <- sapply(
+        datanames,
+        function(name) teal.data::parent(joinkeys, name),
+        USE.NAMES = TRUE,
+        simplify = FALSE
+      )
+      intersect(topological_sort(child_parent), datanames)
```

Comment:
I guess you wanted a `union`?
```suggestion
      union(topological_sort(child_parent), datanames)
```

Reply:
I followed the implementation in `teal.slice` https://github.com/insightsengineering/teal.slice/blob/4366cf9ad01a97210d03e4f776215a758e0b69ff/R/FilteredData.R#L126

This part for not clear, but I guess `topological_sort` might return more names than there are in `datanames`, that's why we limit the result with `intersect`.

Reply:
if `topological_sort(child_parent)` is always a superset of datanames then ok. But union is much safer, because we wan't `datanames + parents` in order where parents preceed their child.

Reply:
Okey, make sense. In case there is a parent in join_keys but not in datanames with union it will appear before the children

Reply:
<img width="357" alt="Skärmavbild 2024-07-31 kl  08 26 17" src="https://github.com/user-attachments/assets/845e2d9a-b199-4602-bb4d-c6170a53a3ca">

From the last tab in the above example

Reply:
Got it @gogonzo . `union` was included in the code

Reply:
Also added a test to https://github.com/insightsengineering/teal/pull/1280/commits/a31ebfbb5187a25c01fbbc23ce8761e34864f27e

Reply:
@m7pr example app I've provided still shows wrong order of datanames. Please fix

Reply:
Does it? The last tab has `c("ANL", "iris", "ADTTE")` specified as datanames. There are only `join_keys` for `"ADSL"` and `"ADTTE"` in `data`. So `ADSL` is added because it is a parent of `ADTTE`. So we end up with what's on your print screen. 

Ahhhh, but I get we would like to see `ADTTE` just after `ADSL` and not far behind at the end? I don't think topological sort does that - it only ensures `ADSL` is somewhere before `ADTTE` - but I will investigate.

Reply:
Ah yeah, I see - `ADSL` should be directly before `ADTTE` and there was a stupid typo https://github.com/insightsengineering/teal/pull/1280/commits/5e7717ae216b6c94a1975b1f76e0c745ad0c8391

Reply:
After 2 fix commits the result is still the same as on your print.screen. Give me some more time before I call you for a review! Stay tuned!

Reply:
Ok, got this sorted out
<img width="102" alt="image" src="https://github.com/user-attachments/assets/1fc955e2-4e6e-48dc-b9b5-b4e7274c3445">


We needed to use `teal_data_datanames` instead of `teal.data::datanames` in `summary-table` https://github.com/insightsengineering/teal/pull/1280/commits/be270a8cb1389f7c3a4b0c1fa53e1e5d60682136